app/service/knowledge_svc.py

```diff
@@ -29,52 +29,68 @@ def _load_module(module_type, module_info):
     async def add_fact(self, fact, constraints=None):
         # Add a new fact to the knowledge service
         if isinstance(fact, Fact):
-            return self.__loaded_knowledge_module._add_fact(fact, constraints)
+            return await self.__loaded_knowledge_module._add_fact(fact, constraints)
 
     async def update_fact(self, criteria, updates):
         # Update an existing fact
-        return self.__loaded_knowledge_module._update_fact(criteria, updates)
+        return await self.__loaded_knowledge_module._update_fact(criteria, updates)
 
     async def get_facts(self, criteria, restrictions=None):
         ###
         # Becomes a powerful function, because it sorts and filters out facts based on
         # input (values, groupings) as well as underlying mechanisms such as fact mutexs
         ###
-        return self.__loaded_knowledge_module._get_facts(criteria, restrictions)
+        return await self.__loaded_knowledge_module._get_facts(criteria, restrictions)
 
     async def delete_fact(self, criteria):
         # Delete existing fact based on provided information
-        return self.__loaded_knowledge_module._delete_fact(criteria)
+        return await self.__loaded_knowledge_module._delete_fact(criteria)
 
     async def get_meta_facts(self, meta_fact=None, agent=None, group=None):
         # Returns the complete set of facts associated with a meta-fact construct
-        return self.__loaded_knowledge_module._get_meta_facts(meta_fact, agent, group)
+        return await self.__loaded_knowledge_module._get_meta_facts(meta_fact, agent, group)
 
     async def get_fact_origin(self, fact):
         ###
         # Retrieve the specific origin of a fact. If it was learned in the current operation, parse through
         # links to identify the host it was discovered on.
         ###
-        return self.__loaded_knowledge_module._get_fact_origin(fact)
+        return await self.__loaded_knowledge_module._get_fact_origin(fact)
+
+    async def check_new_fact(self, fact, listing=None):
```

Comment:
Maybe it's just because I'm coming from `is_new_fact` but I found this name confusing. It doesn't return True if the fact is new like the previous functionality or the method's name suggests; instead it returns True if the fact is already present. I think either the method's name should change or the return value should be flipped.

Reply:
That's a fair point. I'll rename it to something that makes more sense... `check_fact_exists` seems reasonable.